### ISSUE TITLE
convert onnx problem

### ISSUE THREAD

**Issue by wlz987** [retsuh-bqw/SRFormer-Text-Det#17]
can convert to onnx?

这能转吗

**Comment by whale1008** [comment]
我修改了少量代码，转换不报错，但是结果对不上
@wlz987

**Comment by wlz987** [comment]
> 我修改了少量代码，转换不报错，但是结果对不上 @wlz987

我能转成 onnx,但是输入形状不能动态了,很奇怪

**Comment by wlz987** [comment]
@whale1008
可视化正常,但还是有点差异

**Comment by whale1008** [comment]
@wlz987 
佬是直接用torch.onnx.export转换的吗，可以分享转换的简要代码吗，万分感谢

**Comment by wlz987** [comment]
你要是找到了如何解决动态输入的问题,记得告诉我:@whale1008

anchor_transformer_detector.py:

if self.training:
            gt_instances = [x["instances"].to(
                self.device) for x in batched_inputs]

            targets = self.prepare_targets(gt_instances, images.tensor.shape[-2:], images.tensor)
            output = self.srformer(images)

            # compute the loss
            loss_dict = self.criterion(output, targets, gt_instances)
            weight_dict = self.criterion.weight_dict

            return loss_dict
        else:
            output = self.srformer(images)
            if getattr(self, "diy_export_onnx", False):
                return {"pred_logits":output["pred_logits"].contiguous(),"pred_ctrl_points":output["pred_ctrl_points"].contiguous()}


demo.py:
if args.input:
        if os.path.isdir(args.input[0]):
            args.input = [os.path.join(args.input[0], fname)
                          for fname in os.listdir(args.input[0])]
        elif len(args.input) == 1:
            args.input = glob.glob(os.path.expanduser(args.input[0]))
            assert args.input, "The input path(s) was not found"
        for path in tqdm.tqdm(args.input, disable=not args.output):
            # use PIL, to be consistent with evaluation
            img = read_image(path, format="BGR")
            start_time = time.time()
            predictions, visualized_output = demo.run_on_image(img)
            demo.predictor.model.diy_export_onnx = True

            class DiyModel(torch.nn.Module):
                def __init__(self, model):
                    super().__init__()
                    self.model = model

                def forward(self, x):
                    x = x.permute(2, 0, 1)  # [h,w,3] -> [3,h,w]
                    batched_inputs = [{"image": x}]
                    return self.model(batched_inputs)
            diy_model = DiyModel(demo.predictor.model).eval()
            onnx_img = cv2.imread("image.jpg")
            onnx_img = cv2.resize(onnx_img,(1000,660))
            onnx_tensor = torch.from_numpy(onnx_img.astype("float32"))
            diy_model.eval()
            with torch.no_grad():
                out = diy_model(onnx_tensor)
            draw(onnx_img, out["pred_logits"].detach().cpu().numpy(), out["pred_ctrl_points"].detach().cpu().numpy(), "result_torch.jpg")
            torch.onnx.export(
                diy_model,
                onnx_tensor, "model.onnx",
                input_names=["input"],
                dynamic_axes={"input": {0: "hight",1: "width"}},
                opset_version=16,
                export_params=True,
                verbose=False,
            )
            session = ort.InferenceSession('model.onnx', providers=[
                'CPUExecutionProvider'])
            onnx_img = cv2.imread("image.jpg")
            onnx_img = cv2.resize(onnx_img,(1000,660))
            onnx_tensor = torch.from_numpy(onnx_img.astype("float32"))
            onnx_output = session.run(None, {"input": onnx_tensor.detach().cpu().numpy()})
            pred_logits = onnx_output[0]
            pred_ctrl_points = onnx_output[1]
            draw(onnx_img, pred_logits, pred_ctrl_points, "result_onnx.jpg")

            # 转换为torch张量以便比较
            onnx_output_tensor = [torch.tensor(i) for i in onnx_output]

            # 比较输出结果
            for i, j in zip(list(out.values()), onnx_output_tensor):
                diff = torch.abs(i - j)
                print(f"最大差异: {diff.max()}")
                print(f"平均差异: {diff.mean()}")
            exit(0)

**Comment by whale1008** [comment]
@wlz987 
导出时，我是使用的（n，c，h，w）：
image是(1,2,h,w)的tensor
```
dynamic_axes = {'input': {0: 'batch_size', 2: 'height', 3: 'width'}, 'output': {0: 'batch_size'}}
torch.onnx.export(
        traceable_model, 
        (image, ), 
        save_onnx_path, 
        opset_version=16,  # > 13
        do_constant_folding=False,  # True
        input_names=['input'],
        output_names=['output'],
        dynamic_axes=dynamic_axes,
        export_params=True
    )
```
anchor_deformable_transformer.py   line:141
_cur += (H_ * W_)
替换成：_cur = _cur + (H_ * W_)   (我这边替换后就不报错了，否则会报错)

**Comment by wlz987** [comment]
你能动态输入吗:@whale1008

**Comment by whale1008** [comment]
@wlz987 部分算子转换后在onnx中是定值，会报错，但输入是可以动态的

**Comment by whale1008** [comment]
@wlz987 ms_deform_attn.py   line:151，转换时ms_deform改成pytorch的代码，我这边是可以动态输入且结果能对上了
```
        if torch.onnx.is_in_onnx_export():
            output = ms_deform_attn_core_pytorch(value, input_spatial_shapes, sampling_locations, attention_weights)  
        else:
            output = _MSDeformAttnFunction.apply(
                value, input_spatial_shapes, input_level_start_index, sampling_locations, attention_weights, self.im2col_step)
        output = self.output_proj(output)
        return output
```

**Comment by wlz987** [comment]
@whale1008 
![a1cf5b391306bc1c516b34ebb85b651d](https://github.com/user-attachments/assets/09e4615e-2c9f-4ef1-9424-5d991d4bedd2)
![4fc4b5422bc2317d3e33b42b7ce8f471](https://github.com/user-attachments/assets/8be75dd9-a1ea-4ca0-acbb-c9eb1c1b9a8e)
你没有这个问题吗

**Comment by whale1008** [comment]
@wlz987 没有，不过我把输入改成了（b，c，h，w）的形式，不再是list[dict，dict]，不确定是不是因为这个

**Comment by wlz987** [comment]
@whale1008 
你建个仓库咯

**Comment by whale1008** [comment]
@wlz987 我只改了上面那个代码和输入
```
    def preprocess_image(self, batched_inputs):
        """
        Normalize, pad and batch the input images.
        """
        if torch.onnx.is_in_onnx_export():
            return batched_inputs.to("cuda")
        else:
            images = [self.normalizer(x["image"].to(self.device))
                  for x in batched_inputs]

        images = ImageList.from_tensors(images)
        return images
```

输入是这样：
```
        original_image = read_image(image_path, format=cfg.INPUT.FORMAT)  
        aug = T.ResizeShortestEdge(
            [cfg.INPUT.MIN_SIZE_TEST, cfg.INPUT.MIN_SIZE_TEST], cfg.INPUT.MAX_SIZE_TEST
        )
        height, width = original_image.shape[:2]
        print("height, width:", height, width)
        image = aug.get_transform(original_image).apply_image(original_image)
        
        image = torch.from_numpy(image.astype("float32").transpose(2, 0, 1))
        image = image.unsqueeze(0) 
        
        pixel_mean = torch.Tensor(cfg.MODEL.PIXEL_MEAN).to(image.device).view(3, 1, 1)
        pixel_std = torch.Tensor(cfg.MODEL.PIXEL_STD).to(image.device).view(3, 1, 1)
        image = (image - pixel_mean) / pixel_std
        
        torch_model = build_model(self.cfg)
        DetectionCheckpointer(torch_model).resume_or_load(cfg.MODEL.WEIGHTS)
        torch_model.eval()
```

**Comment by wlz987** [comment]
@whale1008 :
```
import argparse
import glob
import multiprocessing as mp
import os
import time
import cv2
import tqdm
import torch
from detectron2.data import transforms as T
from detectron2.checkpoint import DetectionCheckpointer

from detectron2.data.detection_utils import read_image
from adet.config import get_cfg

from detectron2.modeling import build_model

def setup_cfg(args):
    # load config from file and command-line arguments
    cfg = get_cfg()
    cfg.merge_from_file(args.config_file)
    cfg.merge_from_list(args.opts)
    # Set score_threshold for builtin models
    cfg.MODEL.RETINANET.SCORE_THRESH_TEST = args.confidence_threshold
    cfg.MODEL.ROI_HEADS.SCORE_THRESH_TEST = args.confidence_threshold
    cfg.MODEL.FCOS.INFERENCE_TH_TEST = args.confidence_threshold
    cfg.MODEL.MEInst.INFERENCE_TH_TEST = args.confidence_threshold
    cfg.MODEL.PANOPTIC_FPN.COMBINE.INSTANCES_CONFIDENCE_THRESH = args.confidence_threshold
    cfg.freeze()
    return cfg


def get_parser():
    parser = argparse.ArgumentParser(description="Detectron2 Demo")
    parser.add_argument(
        "--config-file",
        default="configs/quick_schedules/e2e_mask_rcnn_R_50_FPN_inference_acc_test.yaml",
        metavar="FILE",
        help="path to config file",
    )
    parser.add_argument("--webcam", action="store_true", help="Take inputs from webcam.")
    parser.add_argument("--video-input", help="Path to video file.")
    parser.add_argument("--input", nargs="+", help="A list of space separated input images")
    parser.add_argument(
        "--output",
        help="A file or directory to save output visualizations. "
        "If not given, will show output in an OpenCV window.",
    )

    parser.add_argument(
        "--confidence-threshold",
        type=float,
        default=0.3,
        help="Minimum score for instance predictions to be shown",
    )
    parser.add_argument(
        "--opts",
        help="Modify config options using the command-line 'KEY VALUE' pairs",
        default=[],
        nargs=argparse.REMAINDER,
    )
    return parser

if __name__ == "__main__":
    mp.set_start_method("spawn", force=True)
    args = get_parser().parse_args()
    cfg = setup_cfg(args)
    
    original_image = read_image(args.input[0], format=cfg.INPUT.FORMAT)  
    aug = T.ResizeShortestEdge(
        [cfg.INPUT.MIN_SIZE_TEST, cfg.INPUT.MIN_SIZE_TEST], cfg.INPUT.MAX_SIZE_TEST
    )
    height, width = original_image.shape[:2]
    print("height, width:", height, width)
    image = aug.get_transform(original_image).apply_image(original_image)

    image = torch.from_numpy(image.astype("float32").transpose(2, 0, 1))
    image = image.unsqueeze(0)
    pixel_mean = torch.Tensor(cfg.MODEL.PIXEL_MEAN).to(image.device).view(3, 1, 1)
    pixel_std = torch.Tensor(cfg.MODEL.PIXEL_STD).to(image.device).view(3, 1, 1)
    image = (image - pixel_mean) / pixel_std

    torch_model = build_model(cfg)
    DetectionCheckpointer(torch_model).resume_or_load(cfg.MODEL.WEIGHTS)
    torch_model.eval()

    torch_model.diy_export_onnx = True
    dynamic_axes = {'input': {0: 'batch_size', 2: 'height', 3: 'width'}, 'output': {0: 'batch_size'}}
    torch.onnx.export(
            torch_model, 
            (image, ), 
            "model.onnx", 
            opset_version=16,  # > 13
            do_constant_folding=False,  # True
            input_names=['input'],
            output_names=['output'],
            dynamic_axes=dynamic_axes,
            export_params=True
        )
    torch_model.diy_export_onnx = False
   ```
   
   ```
       def preprocess_image(self, batched_inputs):
        """
        Normalize, pad and batch the input images.
        """
        if torch.onnx.is_in_onnx_export():
            images = [i for i in batched_inputs.to("cuda")]
        else:
            images = [self.normalizer(x["image"].to(self.device))
                  for x in batched_inputs]

        images = ImageList.from_tensors(images)
        return images
        
  output = self.srformer(images)
            if getattr(self, "diy_export_onnx", False):
                return {"pred_logits":output["pred_logits"].contiguous(),"pred_ctrl_points":output["pred_ctrl_points"].contiguous()}
```
是这样吗,但是还是不能动态输入啊

**Comment by whale1008** [comment]
@wlz987 用netron看下结构呢

**Comment by wlz987** [comment]
![2b83c770fa08ed7e5771be4a9021eaaa_origin](https://github.com/user-attachments/assets/e4a1d04a-6ce2-4768-a7ed-db29be385aaf)
备注邮箱,发送 onnx动态输入形状导出 ,解决方案压缩包提供👀
